### PR TITLE
[Exam] Improve Generate Participation

### PR DESCRIPTION
**Pull request by stefanwaldhauser**
### Motivation and Context

- [X] - #1684 was implemented using `@Transactional`. The goal of this pull request was to replace it with an implementation based on `@EntityGraph`
- [X]  Show Success / Error Message 

### Steps for Testing

Same as #1684 

### Screenshots
image
image
image
image

### CODE REVIEW

File: src/test/java/de/tum/in/www1/artemis/ExamIntegrationTest.java [hunk 1]
```diff
@@ -190,33 +189,46 @@ public void testStartExercisesWithTextExercise() throws Exception {
...
+        assertThat(noGeneratedParticipations).isEqualTo(exam2.getStudentExams().size());
+        List<Participation> studentParticipations = participationRepository.findAll();
+
+        for (Participation participation : studentParticipations) {
             assertThat(participation.getExercise().equals(textExercise));
             assertThat(participation.getExercise().getCourseViaExerciseGroupOrCourseMember() == null);
             assertThat(participation.getExercise().getExerciseGroup() == exam2.getExerciseGroups().get(0));
             // TODO: check that submissions have been created to the participations of text exercises
         }
+        System.out.println("Test");
```
Comment (anditurdiu) [inline src/test/java/de/tum/in/www1/artemis/ExamIntegrationTest.java:224]: This should be deleted :D

**Comment by anditurdiu** [comment]
Code looks good, but testing is blocked until PR #1710 is merged

**Comment by tiffca** [comment]
Followed steps in #1684 and saw some errors, although it could be due to not-yet implemented functionality. 

**Scenario: Generate student exams with an empty mandatory exercise group** 
Expected: throws an error message to the user
Actual: http failure response 
Steps: 
1. Create an exam
2. Create 2 exercise groups, both mandatory 
3. Create multiple exercises in the first exercise group. 
4. Leave the 2nd exercise group empty. 
5. Add students to exam 
6. Go to “student exams” 
7. Click “generate individual exams” 
Error: 
image

Another error message from testing with a different exam: 
image

zone.js:3324 POST https://artemistest.ase.in.tum.de/api/courses/1/exams/86/generate-student-exams 400
instrument.js:110 50 websocket subscriptions:  MapIterator {"/topic/management/feature-toggles", "/topic/user/9872/notifications", "/topic/course/1/INSTRUCTOR", "/topic/course/21/INSTRUCTOR", "/topic/course/28/INSTRUCTOR", …}

If I add an exercise to the other previously empty exercise group, the “generate individual exams” functionality works as expected. 
image
image

Then after clicking “prepare exercise start” 
-	It says 9 exercises successfully prepared  (for 5 students and 2 exercise groups, 1 group with I think less than 5 exercises and the other group with 1 exercise). 

**Scenario: Re-Generate and re-prepare exercise start returns different numbers** 
Expected: returns the same number of generated and prepared exercises
Actual: returns different numbers each time 
Steps: 
1. As an instructor, select the option to "generate individual exams"
2. Re-click on “generate individual exams” and then “prepare exercise start”, and then do it again in that order 3. returns different # exercises generated and prepared
image

I’m not sure if we’re expecting the same results. 

**Scenario: Deleting an exercise in an exercise group after generating an individual exam returns an error** 
Expected: no error or a message is shown to the user
Actual: returns different numbers each time 
Steps: 
1. As an instructor, create an exam with 2 mandatory exercise groups, both with at least 1 exercise in each group
2. As an instructor, select the option to "generate individual exams" and “prepare exercise start”
3. go back to the exercise groups and delete all existing exercise from one of the exercise groups
4. I receive this message after deleting the exercise. 
image
Failed to load resource: the server responded with a status of 500 ()
instrument.js:110 


instrument.js:110 50 websocket subscriptions:  MapIterator {"/topic/management/feature-toggles", "/topic/user/9872/notifications", "/topic/course/1/INSTRUCTOR", "/topic/course/21/INSTRUCTOR", "/topic/course/28/INSTRUCTOR", …}

Network: 
Request URL: https://artemistest.ase.in.tum.de/api/text-exercises/2712
Request Method: DELETE
Status Code: 500

However this could all be expected and due to unimplemented functionality.

**Comment by alexmardale** [comment]
**Exam deletion**
After generating student exams (if there is at least 1 student registered for the exam), I cannot delete the exam anymore.
image

It seems to be possible to delete exams before generating student exams.


**Number of students registered for exam**

On the "Student Exams" page the number of users registered for the exam is returned as 0, even if there are students registered:
image

But the correct number seems to be returned on the "Exam management page":
image



**"Prepare exercise start"**
The response when clicking on "Prepare exercise start" seems to only contain the number of exercises prepared:
image

but not information about them, as shown here:
https://github.com/ls1intum/Artemis/pull/1684#pullrequestreview-434664595

**Comment by julian-christl** [comment]
> After generating student exams (if there is at least 1 student registered for the exam), I cannot delete the exam anymore.
I can't confirm this, working for me.

Everything else I can confirm. I'll wait for the chance to test the fixed before approving.

**Comment by krusche** [comment]
Thanks for the detailed tests. You found a lot of smaller issues that should definitely be fixed in a follow up. However most findings seem unrelated to the changes in this PR or need larger changes.

I would suggest that we address them in another PR and we focus on the functionality when it is executed once in the right order without exceptional use cases such as delete or regenerate

**Comment by pzdr7** [comment]
> I think I was able to fix the issue that exams could not be deleted 

I can confirm that this works now.

**Comment by stefanwaldhauser** [comment]
It is not a bug but a feature that we only send the number of generated participations. I should have made it clear that this is a change in this branch.

**Comment by stefanwaldhauser** [comment]
Fixed the faulty error display

**Pull request merged**